:warning: [Warm Replicas] Implement warm replica support for controllers.

godwinpang opened the pull request:
This change implements the proposal for warm replicas as proposed in #3121.
It adds a `NeedWarmUp` option for controllers to optionally start as warmed replicas.

Note for reviewers: This draft PR is feature complete with tests but the main purpose is to make sure that things are on the right track. Some of the naming / comments are a bit inconsistent, I will address them in a followup cleanup tomorrow.

Builds upon #3190.

sbueringer commented:
/ok-to-test

sbueringer commented:
I would recommend using  the [WIP] PR title prefix instead of draft so prow runs unit tests

alvaroaleman commented on pkg/config/controller.go:
```diff
@@ -60,6 +60,10 @@ type Controller struct {
 
+	// NeedWarmUp indicates whether the controller needs to use warm up.
+	// Defaults to false, which means the controller will not use warm up.
+	NeedWarmUp *bool
```
I would call it something like `EnableWarmup`, not `Need` - The difference to LeaderElection is that LeaderElection is a dependency of the controller, vs warmup is a property of it

alvaroaleman replied:
This was marked as resolved but not addressed I think?

godwinpang replied:
![Screenshot 2025-05-16 at 12 53 16 AM](https://github.com/user-attachments/assets/c79e4c36-03b0-4dfe-9e56-3c5592cd8ff4)
Thanks, I meant to clean this up at the end too. Seems like Github is bugged r.e. comment resolved

godwinpang replied:
https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/89f5479fbb8b5c43bc98489dd0d3788d5f0a1450

godwinpang commented on pkg/manager/internal.go:
```diff
@@ -439,6 +439,11 @@ func (cm *controllerManager) Start(ctx context.Context) (err error) {
 	}
 
+	// Start and wait for sources to start.
+	if err := cm.runnables.Warmup.Start(cm.internalCtx); err != nil {
```
Will change to coming right after the caches, but don't completely understand the significance of having it after vs. before the non-leader election runnables; mind explaining a bit?

godwinpang replied:
> This needs to block until the Warmup has terminated, otherwise we may end up starting the controller before the sources are started, as the check we added only checks if we started to start the sources, not if we finished doing so

Is this a big problem? Replicas will only start the controller after they win leader election, are you saying that in non-leader election cases the behavior of warmup should be that it completely blocks controller startup?

alvaroaleman replied:
> Will change to coming right after the caches, but don't completely understand the significance of having it after vs. before the non-leader election runnables; mind explaining a bit?

Just so it starts asap

> Is this a big problem? Replicas will only start the controller after they win leader election so I don't see an issue in the leader election failover case; are you saying that in non-leader election cases the behavior of warmup should be that it completely blocks controller startup?

Yes. The expectation in the controller is that the sources are fully started before the rest of the controller starts. If we don't ensure this actually finished before we start the controller, we violate that expectation.

Additionally, there may be other cases where ppl implement this interface and it seems surprising to me that we wouldn't let the warmup finish before we start.

If we wanted to be really fancy, we could do this per-runnable (i.E. not block all runnables on `Warmup` having finished, just the ones whose `Warmup` didn't finish) but that is probably more involved.

godwinpang commented:
/retest

godwinpang commented:
/retest

godwinpang commented:
/retest

godwinpang commented:
/retest

godwinpang commented:
/retest

godwinpang commented:
/retest

godwinpang commented:
/retest

godwinpang commented:
/retest

godwinpang commented:
/retest

godwinpang commented on pkg/internal/controller/controller.go:
```diff
@@ -144,6 +167,40 @@ func (c *Controller[request]) NeedLeaderElection() bool {
+// Warmup implements the manager.WarmupRunnable interface.
+func (c *Controller[request]) Warmup(ctx context.Context) error {
+	if c.NeedWarmup == nil || !*c.NeedWarmup {
+		c.didEventSourcesFinishSync.Store(ptr.To(true))
```
because I had a brainfart while writing the UT, sorry!

godwinpang commented on pkg/internal/controller/controller_test.go:
```diff
@@ -1014,6 +1014,82 @@ var _ = Describe("controller", func() {
 		})
 	})
+
+	Describe("Warmup", func() {
```
done

sbueringer commented:
@godwinpang  I'll review once the current changes are pushed

godwinpang commented on pkg/manager/manager.go:
```diff
@@ -314,6 +314,17 @@ type LeaderElectionRunnable interface {
+
+	// WaitForWarmupComplete is a blocking function that waits for the warmup to be completed. It
+	// returns false if it could not successfully finish warmup.
+	WaitForWarmupComplete(context.Context) bool
```
That was the only way I could figure out how to get it to work with how the runnable groups are setup. My reading of the code for runnable groups is that when we [Start](https://github.com/kubernetes-sigs/controller-runtime/blob/b6c5897febe566008678f765ec5a3a1ea04a123a/pkg/manager/runnable_group.go#L132) a runnable group, it starts all the runnables (calling [reconcile](https://github.com/kubernetes-sigs/controller-runtime/blob/b6c5897febe566008678f765ec5a3a1ea04a123a/pkg/manager/runnable_group.go#L185)) then blocks on the runnableChecks before returning. The runnable check is the only mechanism that can block a runnableGroup from prematurely returning.

I copied this `WaitForWarmupComplete` naming + way of blocking from how the cache syncs was implemented [here](https://github.com/kubernetes-sigs/controller-runtime/blob/b6c5897febe566008678f765ec5a3a1ea04a123a/pkg/manager/runnable_group.go#L63-L65), lmk if you have a better way of implementing this.

alvaroaleman replied:
> The runnable check is the only mechanism that can block a runnableGroup from prematurely returning.

Yes, because this use-case is new. It shouldn't be too difficult to also start tracking which runnables have ended and then compare that number to the total in the group.

sbueringer replied:
> Yes, because this use-case is new. It shouldn't be too difficult to also start tracking which runnables have ended and then compare that number to the total in the group.

Just wanted to note that other runnables usually don't end. Start is blocking for them

godwinpang replied:
The way I am reading this thread is that there is no easy / clean way of tracking runnable completion because warmup runnables are a special case of runnables that actually run to completion. (Arguably the runnables with hasCache do this too?)

There are two options
- Leave this as is because there is no better way of doing this _with the existing runnable group setup_, and also this mirrors  Since this is changed to be an unexported interface, it's probably a relatively easy/minor thing to address as a followup?
- Try to implement runnable completion counting mechanism that @alvaroaleman suggested within this PR itself.

My vote is for the first option, just because this change is already sizeable and the suggested change in the second option seems somewhat invasive despite being simple.

sbueringer replied:
I'm slightly in favor of not refactoring runnable groups in this PR if we can avoid it (always has a risk of introducing bugs elsewhere and this code is already quite complicated), but I'll defer to @alvaroaleman

alvaroaleman replied:
Overall I'd prefer to keep it simple with a single method but we can punt on this for this PR given the interface is unexported

sbueringer replied:
Agree simple would be good. Just not sure how much we have to change in runnable groups.

godwinpang commented:
> @godwinpang I'll review once the current changes are pushed (back from PTO and I have finally time now :))

Just in time :) It should be good for a pass now, I've addressed all the comments from @alvaroaleman's first review other than the outstanding one regarding implementing warmup as a blocking runnable group

sbueringer commented on pkg/manager/internal.go:
```diff
@@ -439,6 +439,11 @@ func (cm *controllerManager) Start(ctx context.Context) (err error) {
 		return fmt.Errorf("failed to start other runnables: %w", err)
 	}
 
+	// Start and wait for sources to start.
```
```suggestion
	// Start WarmupRunnables and wait for warmup to complete.
```

godwinpang replied:
Addressed in https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/6bb4616ba3e651c98015a5382549982074f563ac

sbueringer commented on pkg/manager/internal.go:
```diff
@@ -544,6 +549,10 @@ func (cm *controllerManager) engageStopProcedure(stopComplete <-chan struct{}) e
 
+		// Stop the warmup runnables
+		cm.logger.Info("Stopping and waiting for warmup runnables")
+		cm.runnables.Warmup.StopAndWait(cm.shutdownCtx)
```
Just to confirm my understanding. This will:
* cancel the context that was used to call Warmup()
* wait until all Warmup funcs returned

On the happy path where all Warmup funcs completed during controller startup this will do effectively nothing

godwinpang replied:
Yes. In the happy path `StopAndWait` would just block until the waitgroup tracking runnables has counter == 0 which is will because the Warmup funcs have already completed.

In the unhappy path `StopAndWait` would cancel the context and since the controller's implementation of warmup respects the calling context, it early exits without waiting for warmup to finish which functionally makes sense to be despite it only `Stopping` and not really `Waiting`

sbueringer commented on pkg/internal/controller/controller.go:
```diff
@@ -144,6 +167,38 @@ func (c *Controller[request]) NeedLeaderElection() bool {
+}
+
+// DidFinishWarmup implements the manager.WarmupRunnable interface.
+func (c *Controller[request]) DidFinishWarmup(ctx context.Context) bool {
```
Probably dead code, but should we handle the `if c.NeedWarmup == nil || !*c.NeedWarmup` case here in some way specifically?

sbueringer replied:
Hm. Maybe we have to.

Controller will always be added to the Warmup runnable group.

If Warmup is a no-op, DidFinishWarmup will never return, right?

Which means the manager will get stuck when starting the warmup runnable group

alvaroaleman replied:
Yeah we need some way to short-circuit this code if the warmup is not enabled

godwinpang replied:
Yes, this was an oversight. I've added a fix with a UT [here](http://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/667bb0318710617e96f827b5da631c9c774fab23) that will fail without this short circuit.

sbueringer commented on pkg/manager/internal.go:
```diff
@@ -544,6 +549,10 @@ func (cm *controllerManager) engageStopProcedure(stopComplete <-chan struct{}) e
 		cm.runnables.LeaderElection.startOnce.Do(func() {})
 		cm.runnables.LeaderElection.StopAndWait(cm.shutdownCtx)
 
+		// Stop the warmup runnables
```
Should this be done after l.545? (and before we stop leaderelection runnables)

godwinpang replied:
Actually there is something even more confusing here. There is a comment that says to stop the caches *before* the leader election runnables and that it is an important step, but the call is clearly after stopping leader election runnables?...

godwinpang replied:
Ok, ignoring the confusing ordering for the caches...

I think it is ok to keep this here because the warmup runnable is special and blocks before returning, hence you cannot have a leader election runnable that depends on warmup which needs to be stopped before stopping the warmup runnable. There are only three times we can stop the warmup runnable.

1. It hasn't been started yet. In this case it is a no-op and is safe.
2. It was started but _hasn't_ completed yet. In this case because it hasn't completed, the leader election runnables that depend on the warm up cannot have started yet nor can they be started later, so it is safe to stop the warmup runnables before the leader election runnables.
3. It was started and _has_ completed. In this case stopping it won't do anything, so is safe.

Do you agree or am I missing something?

godwinpang replied:
As an aside I think there should be a test to verify startup + shutdown order since it is kind of important -- I am happy to do that in a followup PR later since this code has existed forever without a test, so I don't want to block this PR on it.

sbueringer replied:
I'll get back to this once we finalized other parts of the code :)

sbueringer replied:
I think we have to call this at the beginning of this go routine (l.542) and we have to start an additional go routine to do it.

Basically during Start we let Others / LeaderElection / Warmup runnables race against each other. 

Worst case. We are still in startEventSources and stuck and we want to cancel the contxt of the runnable group (which is done in StopAndWait).

The problem now is we don't know which of the runnables is stuck in startEventSources and we can't assume any ordering. If we are canceling the "wrong" one here we are stuck. Let me make an example based on the current StopAndWait sequence here.

* A Warmup runnable is stuck in startEventSources and will only return when we cancel the context of the Warmup group
* When we call Others.StopAndWait it will wait until all runnables are completed (via r.wg.Wait())
* The Other runnable corresponding to the stuck Warmup runnable won't complete because the Start func is blocked on the mutex that it shares with the Warmup func (as Warmup is still in progress)

I think there are a few variations of this deadlock possible. So as mentioned above I think the only way to make this safe is to do this in an additional go routine and trigger that at the beginning of this goroutine

@alvaroaleman Does this also make sense to you?

alvaroaleman replied:
To add some more context to the above as it took me a moment to think through this, the reason we need the goroutine is that either `Warmup` or `Start` might be holding the mutex and be blocked on the other, hence cancelling one can't wait for the other to be cancelled, it has to happen in parallel.

godwinpang replied:
[ba51d28](https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/ba51d280cee0066899346eeefbbc1436b7372254)

sbueringer commented on pkg/manager/manager.go:
```diff
@@ -314,6 +314,19 @@ type LeaderElectionRunnable interface {
 	NeedLeaderElection() bool
 }
 
+// WarmupRunnable knows if a Runnable should be a warmup runnable. A warmup runnable is a runnable
```
```suggestion
// WarmupRunnable knows if a Runnable requires warmup. A warmup runnable is a runnable
```

sbueringer commented on pkg/manager/manager.go:
```diff
@@ -314,6 +314,19 @@ type LeaderElectionRunnable interface {
 
+// WarmupRunnable knows if a Runnable should be a warmup runnable. A warmup runnable is a runnable
+// that should be run when the manager is started, but before the leader election is acquired.
+// Note: Implementing this interface is only useful when LeaderElection is enabled, as the behavior
```
nit: Yes and no. Often leader election can be enabled/disabled via command line flags. But at this time the interface is already implemented or not.

Maybe slightly reframe to "when LeaderElection can be enabled"?

godwinpang replied:
Addressed in https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/6bb4616ba3e651c98015a5382549982074f563ac

sbueringer commented on pkg/manager/manager.go:
```diff
@@ -314,6 +314,19 @@ type LeaderElectionRunnable interface {
+// Note: Implementing this interface is only useful when LeaderElection is enabled, as the behavior
+// without leaderelection is to run LeaderElectionRunnables immediately.
+type WarmupRunnable interface {
+	// Warmup is the implementation of the warmup runnable.
```
The comment doesn't really say much :)

godwinpang replied:
Changed, let me know if there's anything else you want me to add

sbueringer commented on pkg/manager/manager.go:
```diff
@@ -314,6 +314,19 @@ type LeaderElectionRunnable interface {
+// that should be run when the manager is started, but before the leader election is acquired.
+// Note: Implementing this interface is only useful when LeaderElection is enabled, as the behavior
+// without leaderelection is to run LeaderElectionRunnables immediately.
+type WarmupRunnable interface {
```
Do we want to export this interface?

If we only need it for controller we could probably keep it unexported for now (like "hasCache")

Also allows us refactorings going forward if necessary

godwinpang replied:
Addressed in https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/6bb4616ba3e651c98015a5382549982074f563ac

sbueringer commented on pkg/manager/manager_test.go:
```diff
@@ -1928,6 +1928,75 @@ var _ = Describe("manger.Manager", func() {
+			Expect(m.Start(ctx)).NotTo(HaveOccurred())
+		}()
+
+		By("Waiting for the warmup runnable to be called")
```
I think this does not actually verify the order. Just that all channels are closed.

e.g. the following also works
```go
		By("Verifying the leader election runnable is called after election")
		<-leaderElectionRunnableCalled

		By("Waiting for leader election to be won")
		<-m.Elected()

		By("Waiting for the warmup runnable to be called")
		<-warmupCalled
```

godwinpang replied:
Rewrote test to verify startup order

sbueringer commented on pkg/manager/runnable_group_test.go:
```diff
@@ -224,3 +348,64 @@ var _ = Describe("runnableGroup", func() {
+// warmupRunnableFunc is a helper struct that implements WarmupRunnable
+// for testing purposes.
+type warmupRunnableFunc struct {
+	RunFunc    func(context.Context) error
```
nit: Let's call this StartFunc, also in leaderElectionRunnableFunc & combinedRunnable

sbueringer commented on pkg/manager/runnable_group_test.go:
```diff
@@ -53,6 +53,130 @@ var _ = Describe("runnables", func() {
+		r := newRunnables(defaultBaseContext, errCh)
+		Expect(r.Add(warmupRunnable)).To(Succeed())
+		Expect(r.Warmup.startQueue).To(HaveLen(1))
+		Expect(r.LeaderElection.startQueue).To(HaveLen(1))
```
Let's verify length of r.Others.startQueue in all these test cases as well

sbueringer commented on pkg/manager/runnable_group_test.go:
```diff
@@ -224,3 +348,64 @@ var _ = Describe("runnableGroup", func() {
+	return true
+}
+
+// combinedRunnable implements both WarmupRunnable and LeaderElectionRunnable
```
I would prefer leaderElectionAndWarmupRunnable

godwinpang replied:
Addressed in https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/6bb4616ba3e651c98015a5382549982074f563ac

sbueringer commented:
Mostly minor findings

godwinpang commented on pkg/internal/controller/controller.go:
```diff
@@ -144,6 +167,38 @@ func (c *Controller[request]) NeedLeaderElection() bool {
+	return c.startEventSources(ctx)
+}
+
+// DidFinishWarmup implements the manager.WarmupRunnable interface.
```
Fixed. Tried to make it more robust do a compile time assertion with `var _ manager.WarmupRunnable = &Controller[]{}`, but we can't do that anymore with keeping the interface unexported.

sbueringer replied:
I would have also expected some unit test failure somewhere if this doesn't implement the interface correctly. Are we missing test coverage?

godwinpang replied:
Yes I agree, thanks for pointing this out. I've added a test that exercises manager + controller integration explicitly in 667bb0318710617e96f827b5da631c9c774fab23. The test fails if controller doesn't implement warmupRunnable.

The previous tests only covered
- Manager can handle runnables that implement warmup interface
- Controller's warmup methods behave as expect.

Question - I'm not sure where the best place to put this test is because it implicitly tests manager functionality by assuming startup order. It seems like putting it in controller (current place) is the lesser of two evils as controllers aren't useful without managers, but managers can be useful without controllers.

sbueringer replied:
controller is fine for me

godwinpang commented:
/retest

godwinpang commented on pkg/internal/controller/controller.go:
```diff
@@ -221,13 +276,19 @@ func (c *Controller[request]) Start(ctx context.Context) error {
+	// CAS returns false if value is already true, so early exit since another goroutine must have
+	// called startEventSources previously
+	if !c.didStartEventSources.CompareAndSwap(false, true) {
+		c.LogConstructor(nil).Info("Skipping starting event sources since it was already started")
```
Addressed in https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/6bb4616ba3e651c98015a5382549982074f563ac

godwinpang commented on pkg/internal/controller/controller.go:
```diff
@@ -83,6 +90,16 @@ type Controller[request comparable] struct {
+	// - nil: not finished syncing
+	// - true: finished syncing without error
+	// - false: finished syncing with error
+	didEventSourcesFinishSync atomic.Value
```
Addressed in https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/6bb4616ba3e651c98015a5382549982074f563ac

godwinpang commented on pkg/controller/controller_test.go:
```diff
@@ -474,5 +474,75 @@ var _ = Describe("controller.Controller", func() {
 			Expect(ok).To(BeFalse())
 		})
+
+		It("should set ShouldWarmupWithoutLeadership correctly", func() {
```
Addressed in https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/6bb4616ba3e651c98015a5382549982074f563ac

godwinpang commented on pkg/controller/controller.go:
```diff
@@ -1,5 +1,4 @@
-/*
-Copyright 2018 The Kubernetes Authors.
+/* Copyright 2018 The Kubernetes Authors.
```
Addressed in https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/6bb4616ba3e651c98015a5382549982074f563ac

godwinpang commented on pkg/config/controller.go:
```diff
@@ -60,6 +60,16 @@ type Controller struct {
+	// improves leadership failover time, as the caches will be prepopulated before the controller
+	// transitions to be leader.
+	//
+	// When set to true, the controller will start its sources without transitioning to be leader.
```
Addressed in https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/6bb4616ba3e651c98015a5382549982074f563ac

sbueringer replied:
Thx, btw you don't have to link the commit where it's fixed (less work for you :))

godwinpang commented on pkg/internal/controller/controller.go:
```diff
@@ -144,6 +167,38 @@ func (c *Controller[request]) NeedLeaderElection() bool {
+
+// DidFinishWarmup implements the manager.WarmupRunnable interface.
+func (c *Controller[request]) DidFinishWarmup(ctx context.Context) bool {
+	err := wait.PollUntilContextCancel(ctx, syncedPollPeriod, true, func(ctx context.Context) (bool, error) {
```
https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/54f4fe313a89d88a55598d196f325dc825855337

Clarified and added a UT in the above commit. `WaitForSync` returns `error` and `WaitForWarmupComplete` returns `bool`, so not exactly 1:1 mapping.

`WaitForSync` treats a context cancellation as a success (nil error), so choosing the same approach here of returning true to indicate that the runnable has finished with no error.

sbueringer replied:
> WaitForSync returns error and WaitForWarmupComplete returns bool, so not exactly 1:1 mapping.

Makes sense. Thx!

sbueringer commented:
I'll review again once https://github.com/kubernetes-sigs/controller-runtime/pull/3192#discussion_r2077712360 is resolved, given the potential big impact on the implementation

alvaroaleman commented on pkg/internal/controller/controller.go:
```diff
@@ -144,6 +166,42 @@ func (c *Controller[request]) NeedLeaderElection() bool {
+		case <-ctx.Done():
+			return true
+		case <-ticker.C:
+			didFinishSync, ok := c.didEventSourcesFinishSyncSuccessfully.Load().(*bool)
```
Can we make `didEventSourcesFinishSyncSuccessfully` a channel which we close after having done that so we don't have to poll here?

alvaroaleman commented on pkg/manager/runnable_group.go:
```diff
@@ -65,8 +67,25 @@ func (r *runnables) Add(fn Runnable) error {
-	case LeaderElectionRunnable:
-		if !runnable.NeedLeaderElection() {
+	case warmupRunnable, LeaderElectionRunnable:
+		if warmupRunnable, ok := fn.(warmupRunnable); ok {
```
If the runnable is a leaderelection runnable and has warmup enabled but leaderelection disabled, we will still do the warmup and then _I think_ the call to warmup might race with the call to `Start` as we start them in parallel? it looks to me like calling `Start` first and then `Warmup` will start the sources twice?

While on paper the above sounds like a nonsensical config it might happen if someone globally enables warmup for all their controllers but has one controller that does not require leader election. Ideally, things just work in that case.

sbueringer replied:
What do you think about 
```go
	// Start WarmupRunnables and wait for warmup to complete.
	if err := cm.runnables.Warmup.Start(cm.internalCtx); err != nil {
		return fmt.Errorf("failed to start warmup runnables: %w", err)
	}

	// Start the non-leaderelection Runnables after the cache has synced.
	if err := cm.runnables.Others.Start(cm.internalCtx); err != nil {
		return fmt.Errorf("failed to start other runnables: %w", err)
	}
```

This should solve it, right? As Warmup.Start waits for completion

alvaroaleman replied:
> This should solve it, right? As Warmup.Start waits for completion

Ah you are right, perfect.

This makes me aware of the fact that now that this is not exported anymore we might not need the "wait for finish" logic if we synchronize inside the controller instead, the main reason I asked for that is that if someone outside of c-r implements the interface they would likely expect warmup to finish before being started. So we can do whichever is overall simpler to implement.

godwinpang replied:
> we might not need the "wait for finish" logic

If you are referring to the `WaitForWarmupComplete` method, we still need it because it's how we tell the runnable group to make Warmup.Start wait for completion

godwinpang replied:
> If I see correctly using sync.Once would just be better in general?

yes. I just didn't know sync.Once existed when I wrote the code 😄

godwinpang replied:
Thought about this a bit more. Does ordering it to do warmup before non leader election runnables break backwards compatibility? Previously a manager with non leader election runnables start right after the cache runnables have been started, but now they block on Warmup to complete. 

For controllers this does not matter as they would block on the event source starting anyways, but this breaks in the case where a user has both controllers and some other runnable implemented in user land added to the manager, since now that other runnable has to wait for the controller's source to finish before it can start.

sbueringer replied:
Good point. Going back to the original comment

> If the runnable is a leaderelection runnable and has warmup enabled but leaderelection disabled, we will still do the warmup and then I think the call to warmup might race with the call to Start as we start them in parallel? it looks to me like calling Start first and then Warmup will start the sources twice?

If we are using sync.Once we will never start the sources twice.
As we are not exporting the interface anymore Controller will be the only implementation of warmupRunnable. So it seems safe to rely on that?

godwinpang replied:
Maybe another way of looking at this is that to preserve backwards compatibility we need the following behavior
- If there is an outstanding call to `startEventSources` then all later calls to the same function should block until that outstanding call returns. I *think* sync.Once already guarantees this, but going to add a UT to verify this behavior. 
- Any function that calls `startEventSources` needs to hold `c.mu` because the critical section on L268 modifies `startWatches` outside of the startEventSources call. 

Simplest solution is probably to take the mu lock in the warmup call.

sbueringer replied:
Just realized that I don't know if sync.Once blocks additional callers

sbueringer replied:
> Any function that calls startEventSources needs to hold c.mu because the critical section on L268 modifies startWatches outside of the startEventSources call.

Maybe worth checking if l.268 should be moved into `startEventSources` (not sure if it changes the fact that we have to hold the lock, but it might be a good improvement in any case)

godwinpang replied:
> Just realized that I don't know if sync.Once blocks additional callers until the code within sync.Once is completed

Looks like it does, so that solves 1 problem: https://go.dev/play/p/XAFFCESpUOe

godwinpang replied:
Pushed code with tests regarding the race + clearing out `startWatches`.

godwinpang commented on pkg/config/controller.go:
```diff
@@ -60,6 +60,16 @@ type Controller struct {
+	//
+	// When set to true, the controller will start its sources without transitioning to be leader.
+	// Defaults to false.
+	NeedWarmup *bool
```
Will document based on the ordering suggested here: https://github.com/kubernetes-sigs/controller-runtime/pull/3192#discussion_r2089531957

godwinpang replied:
updated, but also have a question in #3192 so leaving this open

godwinpang commented on pkg/manager/manager_test.go:
```diff
@@ -1928,6 +1928,79 @@ var _ = Describe("manger.Manager", func() {
+		Expect(m.Add(warmupRunnable)).To(Succeed())
+
+		By("Creating a runnable that requires leader election")
+		leaderElectionRunnable := leaderElectionRunnableFunc{
```
Ah got it, that is a bit counter intuitive but I see it in the code [here](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.20.4/pkg/manager/runnable_group.go#L73-L74) now

godwinpang replied:
Removed

godwinpang commented:
A summary of open comments since they are getting a bit hard to track:
- https://github.com/kubernetes-sigs/controller-runtime/pull/3192#discussion_r2089505996 and https://github.com/kubernetes-sigs/controller-runtime/pull/3192#discussion_r2084453321 discuss runnable group startup ordering.
- https://github.com/kubernetes-sigs/controller-runtime/pull/3192#discussion_r2092655893 discusses Warmup needing to be threadsafe with Start because of access to `c.startWatches`.
- https://github.com/kubernetes-sigs/controller-runtime/pull/3192#discussion_r2077712360 discusses making warmupRunnable a single method interface; dependent on https://github.com/kubernetes-sigs/controller-runtime/issues/3221

Everything else is either resolved + addressed, or kept open because they are blocked on the runnable group startup ordering discussion

godwinpang commented:
/retest

alvaroaleman commented on pkg/internal/controller/controller.go:
```diff
@@ -383,6 +441,15 @@ func (c *Controller[request]) updateMetrics(reconcileTime time.Duration) {
+// ensureDidWarmupFinishChanInitialized ensures that the didWarmupFinishChan is initialized. This is needed
+// because controller can directly be created from other packages like controller.Controller, and
+// there is no way for the caller to pass in the chan.
+func (c *Controller[request]) ensureDidWarmupFinishChanInitialized() {
```
Please instead write a `New` func in pkg/internal/controller and do it there so we don't need this code and the locking around it

godwinpang replied:
Good point, didn't realize it was in `internal` so was worried about keeping backwards compatibility

godwinpang replied:
c9a2973

godwinpang replied:
The reason I did not make New return some sort of interface and keep Controller private is because it would create a circular dep between `pkg/internal/controller` and `pkg/controller`.
I would have had to either define a new interface that replicates the ones in `/pkg/controller/controller.go` which adds the burden of keeping them in sync, so think this warning is good enough given that it's an internal pkg?

alvaroaleman commented on pkg/internal/controller/controller.go:
```diff
@@ -83,6 +83,22 @@ type Controller[request comparable] struct {
+
+	// didWarmupFinish is closed when startEventSources returns. It is used to
+	// signal to WaitForWarmupComplete that the event sources have finished syncing.
+	didWarmupFinishChan chan struct{}
+
+	// didWarmupFinishSuccessfully is used to indicate whether the event sources have finished
+	// successfully. If true, the event sources have finished syncing without error. If false, the
+	// event sources have finished syncing but with error.
+	didWarmupFinishSuccessfully atomic.Bool
```
Please make this a `chan error` and buffer it when initializing, so we can use it both to check if it finishes and if there was an issue and don't need the bool:
```suggestion
	didWarmupFinishChan chan error
```

sbueringer replied:
Resolving this one as the conversation moved on :)

alvaroaleman commented on pkg/internal/controller/controller.go:
```diff
@@ -144,6 +166,37 @@ func (c *Controller[request]) NeedLeaderElection() bool {
 }
 
+// Warmup implements the manager.WarmupRunnable interface.
+func (c *Controller[request]) Warmup(ctx context.Context) error {
```
I guess related to the error comment, having this return an `error` is very confusing because it will then get surfaced in the manager rather than in the controller when starting it - I'd prefer it if we had no return here and instead from the `sync.Once` either close `didWarmupFinishChan` on success or put the error in it.

That will reduce this method to just the sync.Once call if warmup is enabled

godwinpang replied:
> it will then get surfaced in the manager rather than in the controller when starting it

QQ: Are you saying here that if warmup failed with error, the controller's Start method should early exit out with that error? Is that the right way of getting the error to be surfaced in the controller when starting it?

godwinpang replied:
> having this return an error is very confusing

Also, to make warmup a runnable I need to have warmup return error. Are you saying that warmup should swallow the error, always return nil, then delay the error to be surfaced until the controller is started?

alvaroaleman replied:
> QQ: Are you saying here that if warmup failed with error, the controller's Start method should early exit out with that error? Is that the right way of getting the error to be surfaced in the controller when starting it?

Yes ideally. Call the sync.Once, block on the err chan, return err or continue.

> Also, to make warmup a runnable I need to have warmup return error. Are you saying that warmup should swallow the error, always return nil, then delay the error to be surfaced until the controller is started?

Either that or just no return and wrap it in the manager so it becomes a runnable

sbueringer replied:
Would this mean that `cm.runnables.Warmup.Start` returns without error even if it fails and the error would only be surfaced after the controller became leader?

alvaroaleman replied:
> Would this mean that cm.runnables.Warmup.Start returns without error even if it fails and the error would only be surfaced after the controller became leader?

Yes

alvaroaleman replied:
@godwinpang @sbueringer and I discussed this and came to the conclusion that it would be preferable to fail in `Warmup` already so we don't need to acquire the leader lease to error out. So I think what we want in the end is:
* Get rid of the `WaitForWarmupComplete`, we don't need it
* Both `Warmup()` and `Start()` will end up calling `startEventSources` which then calls the sync.Once and returns an error if any - Since we have a guarantee that the sync.Once only executes once, this is threadsafe
* Get rid of the `warmupResultChan` as we don't need it

godwinpang replied:
79a7b95

There are a few discussions so it is confusing, but I am pretty sure that reading between the lines we want to keep the warmup runnable returning error so I left it that way in the commit

sbueringer replied:
> but I am pretty sure that reading between the lines we want to keep the warmup runnable returning error so I left it that way in the commit

Yes! At a first glance the commit looks good. I'll do a full review tomorrow

alvaroaleman commented on pkg/internal/controller/controller.go:
```diff
@@ -221,60 +324,71 @@ func (c *Controller[request]) Start(ctx context.Context) error {
-					err := fmt.Errorf("failed to wait for %s caches to sync %v: %w", c.Name, syncingSource, err)
-					log.Error(err, "Could not wait for Cache to sync")
-					sourceStartErrChan <- err
+	var retErr error
```
I would not return the error from `startEventSources` because then the caller has to make sure they hold the lock. Instead, just write from the sync.Once into the err chan

sbueringer replied:
This one should be outdated now, given https://github.com/kubernetes-sigs/controller-runtime/pull/3192#discussion_r2098142231

alvaroaleman commented on pkg/internal/controller/controller.go:
```diff
@@ -144,6 +220,39 @@ func (c *Controller[request]) NeedLeaderElection() bool {
+
+// WaitForWarmupComplete returns true if warmup has completed without error, and false if there was
+// an error during warmup. If context is cancelled, it returns true.
+func (c *Controller[request]) WaitForWarmupComplete(ctx context.Context) bool {
```
THere is no need for this anymore now I think?

sbueringer replied:
What do you mean with "this" the whole func?

sbueringer replied:
Discussed. The idea is to get rid of the entire func. Let's resolve to dedup with: https://github.com/kubernetes-sigs/controller-runtime/pull/3192#discussion_r2098142231

sbueringer commented on pkg/config/controller.go:
```diff
@@ -60,6 +60,19 @@ type Controller struct {
+	//
+	// Setting EnableWarmup to true and NeedLeaderElection to true means the controller will start its
+	// sources without waiting to become leader.
+	// Setting EnableWarmup to true and NeedLeaderElection is false is a no-op as controllers without
```
```suggestion
	// Setting EnableWarmup to true and NeedLeaderElection to false is a no-op as controllers without
```

sbueringer commented on pkg/controller/controller.go:
```diff
@@ -93,6 +93,16 @@ type TypedOptions[request comparable] struct {
 	// Note: This flag is disabled by default until a future version. It's currently in beta.
 	UsePriorityQueue *bool
+
+	// EnableWarmup specifies whether the controller should start its sources when the manager is not
```
Maybe sync this comment with the one in pkg/config/controller.go (as this field is also exported)

sbueringer commented on pkg/internal/controller/controller.go:
```diff
@@ -38,7 +38,54 @@ import (
+
+	// EnableWarmup specifies whether the controller should start its sources
+	// when the manager is not the leader.
+	// Defaults to false, which means that the controller will wait for leader election to start
```
```suggestion
	// Defaults to false, which means that the controller will wait for leader election
```
or ~ sync with the other fields

sbueringer commented on pkg/internal/controller/controller.go:
```diff
@@ -83,6 +130,13 @@ type Controller[request comparable] struct {
+	didStartEventSourcesOnce sync.Once
+
+	// warmupResultChan receives the result (nil / non-nil error) of the warmup method. It is
+	// consumed by the WaitForWarmupComplete method that the warmup has finished.
```
I think there's something wrong or missing in this sentence

godwinpang replied:
removed that field altogether :)

sbueringer commented on pkg/internal/controller/controller.go:
```diff
@@ -144,6 +220,39 @@ func (c *Controller[request]) NeedLeaderElection() bool {
+		return false
+	}
+
+	return warmupError == nil
```
So what happens if we hit the error case?

Seems like we then never signal ready on startReadyCh, which means the runnableGroup.Start func only returns if the ctx is cancelled? (which if I see correctly is equivalent to it only returns if the ctx passed into controllerManager.Start returns?)

sbueringer reviewed (commented):
Only had enough time to review the prod code. Wil review tests on the next round. I hope I found all the issues :)

sbueringer commented on pkg/internal/controller/controller_test.go:
```diff
@@ -1014,6 +1103,207 @@ var _ = Describe("controller", func() {
+			const warmupRunnableName = "warmupRunnable"
+
+			ctrl.CacheSyncTimeout = time.Second
+			ctrl.startWatches = []source.TypedSource[reconcile.Request]{
```
What makes this a warmupRunnable?

godwinpang replied:
The JustBeforeEach block on L1108 sets `warmup == true`

sbueringer replied:
Maybe let's make this explicit

sbueringer commented on pkg/internal/controller/controller_test.go:
```diff
@@ -1014,6 +1103,207 @@ var _ = Describe("controller", func() {
+			}()
+
+			<-m.Elected()
+			Expect(<-runnableExecutionOrderChan).To(Equal(warmupRunnableName))
+			Expect(<-runnableExecutionOrderChan).To(Equal(nonWarmupRunnableName))
```
What part of the prod code guarantees this behavior?

1. We set LeaderElection on the manager to false
2. Warmup & Start funcs currently should race against each other afaik?

godwinpang replied:
Setting LeaderElection on the manager == false was a mistake, will fix.

sbueringer commented on pkg/internal/controller/controller_test.go:
```diff
@@ -1014,6 +1103,207 @@ var _ = Describe("controller", func() {
+			Expect(<-runnableExecutionOrderChan).To(Equal(nonWarmupRunnableName))
+		})
+
+		It("should not race with Start and only start sources once", func() {
```
is the first part true? I mean Start and Warmup are racing (but "only start sources once" is true)

godwinpang replied:
By `race`, I meant `should not called a race condition when called concurrently with Start`. Let me reword to make it clear

sbueringer commented on pkg/internal/controller/controller_test.go:
```diff
@@ -1014,6 +1103,207 @@ var _ = Describe("controller", func() {
+
+			By("Calling Start when EnableWarmup is false")
+			// Now call Start
+			go func() {
```
Do we need the go routine?

alvaroaleman replied:
I think so, because `Start` is blocking, one of that or the check needs to be in a goroutine

sbueringer replied:
Ah yes, you're right

sbueringer reviewed (commented):
Also reviewed tests now

alvaroaleman commented on pkg/internal/controller/controller.go:
```diff
@@ -221,60 +299,71 @@ func (c *Controller[request]) Start(ctx context.Context) error {
 // startEventSources launches all the sources registered with this controller and waits
 // for them to sync. It returns an error if any of the sources fail to start or sync.
 func (c *Controller[request]) startEventSources(ctx context.Context) error {
```
Yes, makes sense. Also, lets add a test for this

godwinpang replied:
https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/ba51d280cee0066899346eeefbbc1436b7372254 is my stab at it but there is another race that this uncovers. This is also why CI is failing.

[L336](https://github.com/kubernetes-sigs/controller-runtime/pull/3192/files#diff-3f86588ef108e53ef346d3223ca106ccaa8efe72cb1055489665d7675f3da7b4R336-R339) starts a goroutine that can potentially outlive the duration of the caller holding the lock. The errGroup blocks on `case <-sourceStartCtx.Done()`, meaning that when the context is cancelled, startEventSources() doesn't wait for the the goroutine on [L336](https://github.com/kubernetes-sigs/controller-runtime/pull/3192/files#diff-3f86588ef108e53ef346d3223ca106ccaa8efe72cb1055489665d7675f3da7b4R336-R339) to complete before returning. This means that the mutex acquired in Warmup can be released before [L339](https://github.com/kubernetes-sigs/controller-runtime/pull/3192/files#diff-3f86588ef108e53ef346d3223ca106ccaa8efe72cb1055489665d7675f3da7b4R336-R339) is executed.

The problem variable is `c.Queue` since it can be read on [L339](https://github.com/kubernetes-sigs/controller-runtime/pull/3192/files#diff-3f86588ef108e53ef346d3223ca106ccaa8efe72cb1055489665d7675f3da7b4R336-R339), but also assigned to in Start on [L261](https://github.com/kubernetes-sigs/controller-runtime/pull/3192/files#diff-3f86588ef108e53ef346d3223ca106ccaa8efe72cb1055489665d7675f3da7b4R261). 

It seems *reasonable* to ensure that the `watch.Start` call executes before the errgroup returns in the context cancelled case, as the `watch.Start` ctx call should early exit when called with a cancelled / expired context.

I don't have a solution just yet since it is late, but rough ideas are to make didStartSyncingSource` a channel instead and block on it in the ctx cancelled case before proceeding with error handling. The one problem with that is that the error message on L365 would be dropped and be backwards incompatible

godwinpang replied:
Maybe we can do something as follows instead? Keep the current error handling behavior, but add a channel blocking until watch.Start is called in the defer block
```
sourceStartErrChan := make(chan error, 1) // Buffer chan to not leak goroutine if we time out
didStartSyncingSourceChan := make(chan struct{}) // <-- ⏰ this is the change
go func() {
	defer close(sourceStartErrChan)
	log.Info("Starting EventSource")
	if err := watch.Start(ctx, c.Queue); err != nil {
		sourceStartErrChan <- err
		return
	}
        close(didStartSyncingSourceChan) // <-- ⏰ this is the change
	syncingSource, ok := watch.(source.TypedSyncingSource[request])
	if !ok {
		return
	}
	didStartSyncingSource.Store(true)
	if err := syncingSource.WaitForSync(sourceStartCtx); err != nil {
		err := fmt.Errorf("failed to wait for %s caches to sync %v: %w", c.Name, syncingSource, err)
		log.Error(err, "Could not wait for Cache to sync")
		sourceStartErrChan <- err
	}
}()

select {
defer func() {<-didStartSyncingSourceChan}() // <-- ⏰ this is the change
case err := <-sourceStartErrChan:
	return err
case <-sourceStartCtx.Done():
	if didStartSyncingSource.Load() { // We are racing with WaitForSync, wait for it to let it tell us what happened
		return <-sourceStartErrChan
	}
	if ctx.Err() != nil { // Don't return an error if the root context got cancelled
		return nil
	}
	return fmt.Errorf("timed out waiting for source %s to Start. Please ensure that its Start() method is non-blocking", watch)
}
```

IDT we can overload `sourceStartErrChan` because we would have to do something like produce a nil value to it right after `watch.Start(ctx, c.Queue)` which breaks the error capturing case in WaitForSync.
It wouldn't work to overload `didStartSyncingSource` either since we'd have to poll it until it turns true.

sbueringer replied:
I think the problem is bigger. Does Warmup start the sources with a nil queue? (if it is called before Start)

sbueringer replied:
I think we would have to write c.Queue already in New. This also starts the queue though.

So this would also change the observable behavior of functions like NewUnmanaged / NewTypedUnmanaged.

So I think we can't move it into New. 

The easiest might be to move it into startEventSourcesLocked (and rename that func to startEventSourcesAndQueueLocked)

alvaroaleman replied:
Can we make sure that our tests validate that the queue we pass to the sources is non-nil? This should've been caught by tests and not by a human ideally

godwinpang replied:
I added a UT to verify this in [12e938c](https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/12e938c905595368c7ebfca5462fce673be73fb6) and also changed the existing integration tests at both the manager and controller levels to exercise controllers with warmup enabled so there is some e2e coverage.

Let me also go through the existing tests around `Start`  in [controller_test.go](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/internal/controller/controller_test.go) and replicate them with warmup to make sure that we have coverage for existing tested behavior too.

godwinpang commented on pkg/internal/controller/controller.go:
```diff
@@ -144,6 +219,15 @@ func (c *Controller[request]) NeedLeaderElection() bool {
+		return nil
+	}
+
+	return c.startEventSources(ctx)
```
[5df573f](https://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/5df573fde37b68239e45503a12e9afeac2ec7ce4)

sbueringer commented:
@godwinpang Pleae keep the conversations open. It's really hard to find the just resolved/fixed conversations to check that everything was addressed

sbueringer commented on pkg/internal/controller/controller.go:
```diff
@@ -218,63 +308,83 @@ func (c *Controller[request]) Start(ctx context.Context) error {
+		// which won't be garbage collected if we hold a reference to it.
+		c.startWatches = nil
+
+		// Mark event sources as started after resetting the startWatches slice to no-op a Watch()
```
Watch won't be a no-op. It will start the source immediately instead of storing it in startWatches

sbueringer commented:
@godwinpang Heyho, discussed this with @alvaroaleman.

I think we're making really good progress on this PR and getting very close to merging it. We would just like to take a bit more time to ensure we don't break anything (I"ll also try to test it with Cluster API).

So we'll go ahead and cut the CR v0.21 release now already (a few folks are waiting for that :)). Once this PR merges we'll cherry-pick and cut a v0.21.1 patch release.

(I also modified the PR title as this PR is not breaking)

alvaroaleman commented:
@godwinpang any plans on wrapping this up?

godwinpang commented:
> @godwinpang any plans on wrapping this up?

Sorry, I missed this Github notification. I've updated the PR description and also duplicated the tests between `Start` and `Warmup` as appropriate.

alvaroaleman reviewed (approved):
thanks!
/hold

in case Stefan has anything to add

godwinpang commented on pkg/internal/controller/controller.go:
```diff
@@ -218,63 +297,94 @@ func (c *Controller[request]) Start(ctx context.Context) error {
+				case err := <-sourceStartErrChan:
+					return err
+				case <-sourceStartCtx.Done():
+					defer func() { <-hasAccessedQueueChan }() // Ensure that watch.Start has been called to avoid prematurely releasing lock before accessing c.Queue
```
yes, I agree we don't need this synchronization anymore with the move into sync.Once. I'll take out the code that protects against races on `c.Queue` access

godwinpang commented on pkg/controller/controller_integration_test.go:
```diff
@@ -48,30 +53,46 @@ var _ = Describe("controller", func() {
 		// TODO(directxman12): write a whole suite of controller-client interaction tests
 
-		It("should reconcile", func() {
+		// Since all tests are run in parallel and share the same testenv, we namespace the objects
```
They're not. I was confused because when I added the tests pods from the other namespace started showing up in the reconciled channel. I checked again and looks like the main reason is that the pod isn't deleted in the integration tests unlike the deployment / replicasets.

I tried adding a delete call to the test, but tests would start consistently failing with the `timeout waiting for process kube-apiserver to stop` error which is described in https://github.com/kubernetes-sigs/controller-runtime/issues/1571. I didn't want to add a hard coded time.Sleep described in the issue, so I would propose keeping this namespace predicate as a workaround to isolate tests. (I'll remove the running tests in parallel comment)

sbueringer replied:
Sounds good. 

> I'll remove the running tests in parallel comment

Please add the new explanation instead :)

godwinpang commented on pkg/manager/manager_test.go:
```diff
@@ -1928,6 +1928,64 @@ var _ = Describe("manger.Manager", func() {
+		}()
+
+		<-m.Elected()
+		Expect(<-runnableExecutionOrderChan).To(Equal(warmupRunnableName))
```
Right, it doesn't. It seems like there isn't a way to manually override / control leader election behavior in tests, so going to add a leader election fake and add it to this

sbueringer commented on pkg/internal/controller/controller.go:
```diff
@@ -218,63 +297,91 @@ func (c *Controller[request]) Start(ctx context.Context) error {
+					defer close(sourceStartErrChan)
+					log.Info("Starting EventSource")
+
+					q := c.Queue
+					if err := watch.Start(ctx, q); err != nil {
```
```suggestion
					if err := watch.Start(ctx, c.Queue); err != nil {
```

sbueringer commented on pkg/manager/manager_test.go:
```diff
@@ -1928,6 +1928,79 @@ var _ = Describe("manger.Manager", func() {
+		By("Unblocking leader election")
+		resourceLockWithHooks.UnblockLeaderElection()
+
+		By("Waiting for the leader election runnable to be executed without leader election being won")
```
```suggestion
		By("Waiting for the leader election runnable to be executed after leader election was won")
```

sbueringer commented on pkg/leaderelection/fake/leader_election.go:
```diff
@@ -30,6 +32,20 @@ import (
+
+	// BlockLeaderElection blocks the leader election process when called. It will not be unblocked
+	// until UnblockLeaderElection is called.
+	// Not thread safe.
```
nit: I'm not sure I follow what makes this not thread safe (same in l.46)

Or maybe the comment is outdated?

sbueringer reviewed (commented):
Thx, just some very minor findings now.

@alvaroaleman Can you please also take another look?

godwinpang commented:
/retest

sbueringer commented:
Thank you!

/lgtm

/assign @alvaroaleman Please take a final look, thx!

Once this PR merges, let's please update the proposal PR (#3121) and merge it as well
